feat: library home page bare bones [FC-0059]

rpenido opened the pull request:
## Description
This PR adds a "Bare Bones" version of the Library Home and updates the SearchResult component to redirect to this new page.

![image](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/0611fdfa-e357-471d-9561-7ea0ee0a696a)

![image](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/e1bf1807-6a02-47e3-a268-38d210897687)

## More Info
- Closes: https://github.com/openedx/frontend-app-course-authoring/issues/1036
- Depends on: https://github.com/openedx/frontend-app-course-authoring/pull/1050

## Testing Instructions
- Run your local stack on this branch
- Make sure you have the studio.library_authoring_mfe waffle flag **DISABLED**
- Go to the Studio Home and check the "Libraries" tab
- Click on one library with components
- Check if the component count is shown in the last card and the "Components" tab
- Use the keyword search box and check if the component count is filtered accordingly
- Navigate through the Home, Components and Collections tab and check if you can navigate back using the "Back" button from the brownser
- Go back to the Studio Home and open a library WITHOUT components
- You should see a message with a button that will allow the user to add new components (not implemented yet)
- Go back to the Studio Home and open the search modal
- Find a library v2 component and click on it. It should redirect to the new Library Home.

---
Private ref: [FAL-3753](https://tasks.opencraft.com/browse/FAL-3753)

**Settings**

```yaml
EDX_PLATFORM_REPOSITORY: https://github.com/open-craft/edx-platform.git
EDX_PLATFORM_VERSION: taxonomy-sandbox-20240614

GROVE_REDIRECTS:
 - domain: {{ LMS_HOST }}
 - domain: cms.{{ LMS_HOST }}
 - domain: app.{{ LMS_HOST }}
 - domain: meilsearch.{{ LMS_HOST }}

PLUGINS:
- mfe
- grove
- s3
- meilisearch
```

**Tutor requirements**

```txt
git+https://github.com/overhangio/tutor.git@nightly
git+https://github.com/overhangio/tutor-discovery.git@nightly
git+https://github.com/overhangio/tutor-ecommerce.git@nightly
git+https://github.com/overhangio/tutor-xqueue.git@nightly
git+https://github.com/overhangio/tutor-forum.git@nightly
git+https://gitlab.com/opencraft/dev/tutor-contrib-grove.git@main

git+https://github.com/open-craft/tutor-mfe.git@7d1e1e0dad2c0e4247a4fd8280c257f13ed28432
git+https://github.com/open-craft/tutor-contrib-s3.git@04c7491442fa231e79840995f1f8707eed6353fd
```

openedx-webhooks commented:
Thanks for the pull request, @rpenido! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

rpenido commented on src/CourseAuthoringPage.jsx:
```diff
@@ -15,29 +15,6 @@ import { getCourseAppsApiStatus } from './pages-and-resources/data/selectors';
 import { RequestStatus } from './data/constants';
 import Loading from './generic/Loading';
 
-const AppHeader = ({
-  courseNumber, courseOrg, courseTitle, courseId,
-}) => (
-  <Header
-    courseNumber={courseNumber}
-    courseOrg={courseOrg}
-    courseTitle={courseTitle}
-    courseId={courseId}
-  />
-);
-
-AppHeader.propTypes = {
-  courseId: PropTypes.string.isRequired,
-  courseNumber: PropTypes.string,
-  courseOrg: PropTypes.string,
-  courseTitle: PropTypes.string.isRequired,
-};
-
-AppHeader.defaultProps = {
-  courseNumber: null,
-  courseOrg: null,
-};
-
```
This component was just passing parameters to `<Header />`.

rpenido commented on src/header/Header.jsx:
```diff
@@ -6,57 +6,58 @@ import { useIntl } from '@edx/frontend-platform/i18n';
-  courseOrg,
-  courseNumber,
-  courseTitle,
+  contentId,
+  org,
+  number,
+  title,
   isHiddenMainMenu,
+  isLibrary,
 }) => {
```
We are using the same Header for courses and libraries now.

rpenido commented on src/search-modal/data/apiHooks.js:
```diff
@@ -45,6 +45,9 @@ export const useContentSearchResults = ({
   blockTypesFilter,
   tagsFilter,
 }) => {
+  blockTypesFilter ??= []; // eslint-disable-line no-param-reassign -- default value for optional parameter
+  tagsFilter ??= []; // eslint-disable-line no-param-reassign -- Default value for optional parameter
+
```
Making these parameters optional and passing a default

rpenido commented on src/library-authoring/data/apiHook.js:
```diff
@@ -0,0 +1,52 @@
+  })
+);
+
+/**
+ * Hook to fetch the count of components and collections in a library.
+ * @param {string} libraryId - The ID of the library to fetch.
+ * @param {string} searchKeywords - Keywords to search for.
+ */
+export const useLibraryComponentCount = (libraryId, searchKeywords) => {
+  // Meilisearch code to get Collection and Component counts
+  const { data: connectionDetails } = useContentSearchConnection();
+
+  const indexName = connectionDetails?.indexName;
+  const client = React.useMemo(() => {
+    if (connectionDetails?.apiKey === undefined || connectionDetails?.url === undefined) {
+      return undefined;
+    }
+    return new MeiliSearch({ host: connectionDetails.url, apiKey: connectionDetails.apiKey });
+  }, [connectionDetails?.apiKey, connectionDetails?.url]);
+
+  const libFilter = `context_key = "${libraryId}"`;
+
+  const { totalHits: componentCount } = useContentSearchResults({
+    client,
+    indexName,
+    searchKeywords,
+    extraFilter: [libFilter], // ToDo: Add filter for components when collection is implemented
+  });
+
+  const collectionCount = 0; // ToDo: Implement collections count
+
+  return {
+    componentCount,
+    collectionCount,
+  };
+};
```
I created this hook to get only the component count for the library. Maybe we should replace this with another instance of the `SearchContext` when we add other filters. @ChrisChV @bradenmacdonald

bradenmacdonald replied:
Yes, when we get there what I want to do is move most of the "generic" search code out of `search-modal` into somewhere like `generic/search/` and then we can use it in both `search-modal/SearchModal.tsx` and `src/library-authoring/LibraryHome.tsx` etc.

rpenido replied:
@bradenmacdonald
I was thinking of exporting this directly from our `search-modal` feature. I know the current name is not the best, but it would be our `search-engine` feature that exports both the modal and the provider to use the search engine elsewhere. Maybe rename this feature to `search-engine` to make it explicit.

Or do you think having two features (`search-modal` and `search-engine`) would work best?

rpenido replied:
And, just to confirm: we are moving to use the search index as the primary source for listing content, right?

We might need to move some indexing tasks to run sync because if we add or remove a library, we could refetch before the index update and end up with out of sync data.

bradenmacdonald replied:
Yes, exactly. When we make changes to libraries / library content, we can update the search index synchronously.

It's fine to export from search-modal for now, but I do think it's better to separate the library (search manager) from the two use cases: library home UI, and search modal.

yusuf-musleh commented:
@rpenido I left a comment on the last commit [here](https://github.com/open-craft/frontend-app-course-authoring/commit/c63bc2fbf586a791770b3b63e3262c5c6bd542ec#commitcomment-143301065)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-1076-11bdb6.staging.do.opencraft.hosting)
📝 [Studio](https://studio.pr-1076-11bdb6.staging.do.opencraft.hosting)
ℹ️ [Grove Config](https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-34602668-7137341364-tutor-config.yml), [Tutor Config](https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-34602668-7137341364-grove-config.yml), [Tutor Requirements](https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-34602668-7137341364-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-1076-11bdb6.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-1076-11bdb6.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-51564656-7137497377-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-51564656-7137497377-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-51564656-7137497377-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-1076-11bdb6.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-1076-11bdb6.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-51564656-7139691772-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-51564656-7139691772-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-51564656-7139691772-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-1076-11bdb6.staging.do.opencraft.hosting)
📝 [Studio](https://studio.pr-1076-11bdb6.staging.do.opencraft.hosting)
ℹ️ [Grove Config](https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-34602668-7139705419-tutor-config.yml), [Tutor Config](https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-34602668-7139705419-grove-config.yml), [Tutor Requirements](https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/pr-1076-11bdb6-34602668-7139705419-tutor-requirements.txt)

bradenmacdonald commented on src/header/Header.jsx:
```diff
@@ -6,57 +6,58 @@ import { useIntl } from '@edx/frontend-platform/i18n';
         <SearchModal
           isOpen={isShowSearchModalOpen}
-          courseId={courseId}
+          courseId={isLibrary ? undefined : contentId}
```
For now, we want the _modal_ to only search courses and we're not going to show "Search all Courses and Libraries" even though it's possible.

To search libraries, users will have to click on a specific library and use the new search UI we're building, which is not in a modal.

rpenido replied:
Ack! @bradenmacdonald 
> For now, we want the modal to only search courses and we're not going to show "Search all Courses and Libraries" even though it's possible.

It is worth to mention that currently we show library blocks in the modal and redirect to the parent library if you click on a hit.

Would you like me to include a fix in this PR to limit it only to courses?

bradenmacdonald replied:
Right, yes please - let's keep that functionality in the code but for now add an additional filter to limit the search modal results to courses. Thanks.

yusuf-musleh reviewed (approved):
:+1: @rpenido Looks good to me! Great work!

- [X] I tested this: I followed the testing instructions in the PR, and everything is working as expected.
- [X] I read through the code
- [X] I checked for accessibility issues
- [ ] ~Includes documentation~

xitij2000 commented on src/library-authoring/EmptyStates.tsx:
```diff
@@ -0,0 +1,23 @@
+
+export const NoComponents = () => (
+  <Stack direction="horizontal" gap={3} className="mt-6 justify-content-center">
+    <FormattedMessage {...messages.noComponents} />
+    <Button iconBefore={Add}>
+      <FormattedMessage {...messages.addComponent} />
```
This isn't specific to this code, but I've seen this pattern a lot and I don't personally see the value in it. 

To me the advantage of using FormattedMessage is that you can see the actual text in the react code instead of looking at the messages file. In this case it's better to use `intl.formatMessage` instead. `FormattedMessage` just uses `useIntl` internally anyway.

bradenmacdonald replied:
Are you saying to write `{ intl.formatMessage(messages.addComponent) }` ? That's totally fine, but requires changing this function to have a statement body to be able to include `const intl = useIntl()` in scope, so writing it this way uses one less line of code, and I think is just as simple.

xitij2000 commented on src/library-authoring/LibraryAuthoringPage.tsx:
```diff
@@ -0,0 +1,131 @@
+import { useContentLibrary } from './data/apiHook';
+import messages from './messages';
+
+const TAB_LIST = {
+  home: '',
+  components: 'components',
+  collections: 'collections',
+};
+
```
Since we're using TypeScript now, this could be an enum?

rpenido replied:
Great! Done: 3b2a6ada12210ba15a716d3e302884565dc4b998

xitij2000 reviewed (approved):
Just a few minor nits other than that it's good to go!

bradenmacdonald commented on src/header/Header.tsx:
```diff
@@ -1,83 +1,76 @@
 import messages from './messages';
 
+interface HeaderProps {
+  contentId?: string,
```
```suggestion
  contextId?: string,
```
The generic term for "course or library" is "learning context" so for consistency this parameter should be `contextId` not `contentId`

rpenido replied:
Done: c92b6b6c88950f7b110c7c2e7849bdec28b9621f

bradenmacdonald commented on src/header/Header.tsx:
```diff
@@ -1,83 +1,76 @@
         mainMenuDropdowns={mainMenuDropdowns}
         outlineLink={outlineLink}
-        searchButtonAction={meiliSearchEnabled && openSearchModal}
+        searchButtonAction={meiliSearchEnabled ? openSearchModal : undefined}
```
Let's hide the search button for libraries please.

bradenmacdonald replied:
Well I think people will be confused that the search icon on the library pages doesn't search libraries. But we can leave it for now and see what the product/UX folks say later when they review and try it out on the sandbox. It's a bit of an issue that the UX for searching libraries and the UX for searching courses are so completely different.

bradenmacdonald commented:
This is a weird bug, but if I go to the "Libraries" tab, then click on any library, then click back, the "New Library" button disappears.

<img width="1475" alt="Screenshot" src="https://github.com/openedx/frontend-app-course-authoring/assets/945577/02d9a8d0-755a-4ab3-a5d1-24d41cbe020f">

bradenmacdonald commented on src/studio-home/StudioHome.jsx:
```diff
@@ -106,7 +111,7 @@ const StudioHome = ({ intl }) => {
 
     return headerButtons;
-  }
+  }, [location]);
```
When we add the actual create form, let's make it so this link uses React Router. Right now it causes a full page load, which is unnecessary.

rpenido replied:
Fixed here: 4df024ecd1211a10638f40739a45f33b7b18bca7

bradenmacdonald commented on src/library-authoring/LibraryAuthoringPage.test.tsx:
```diff
@@ -0,0 +1,236 @@
+  return mockEmptyResult;
+};
+
+const libraryData = {
```
```suggestion
const libraryData: ContentLibrary = {
```

and change the import above to `import { getContentLibraryApiUrl, type ContentLibrary } from './data/api';`

Because you have some type errors here, like `allowPublic_learning` should be `allowPublicLearning`

rpenido replied:
Fixed here: 4d9da7f4e02935e0316bcce33ca84b75dce2bcf5

bradenmacdonald commented on src/library-authoring/LibraryHome.tsx:
```diff
@@ -0,0 +1,54 @@
+
+  return (
+    <Stack gap={3}>
+      <Section title="Recently Modified">
```
These titles need i18n

rpenido replied:
Fixed: 2a63186860cc87aa3d6e4389562ef2a0f261f464

bradenmacdonald commented on src/library-authoring/data/apiHook.ts:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Hook to fetch the count of components and collections in a library.
+ */
+export const useLibraryComponentCount = (libraryId: string, searchKeywords: string) => {
```
This is temporary right? In the future we should have a `<SearchContextProvider>` on the Library Home page, and there should be no need to initialize a Meilisearch client manually.

rpenido replied:
Yes! This was refactored here: https://github.com/openedx/frontend-app-course-authoring/pull/1141

bradenmacdonald commented on src/search-modal/index.js:
```diff
@@ -0,0 +1,3 @@
+// @ts-check
```
Just make this file `.ts` please. Or do we really need this file?

rpenido replied:
I think this is good for helping us avoid inter-feature import. We should not import from `./seach-modal/Something`. We should always import from explicitly exported members from `./search-modal`

rpenido commented on src/library-authoring/LibraryAuthoringPage.test.tsx:
```diff
@@ -0,0 +1,236 @@
+      </QueryClientProvider>
+    </IntlProvider>
+  </AppProvider>
+);
```
Thank you @bradenmacdonald . I was a bit divided about this before.
This structure explicitly states what is needed for this particular component, but it also adds a lot of boilerplate to the tests. 

For the next tests, I will add some helpers for this (and the meilisearch stuff).

rpenido commented on src/library-authoring/LibraryAuthoringPage.tsx:
```diff
@@ -0,0 +1,131 @@
+    } else {
+      setTabKey(TAB_LIST.home);
+    }
+  }, [location]);
```
Great suggestion @bradenmacdonald! Fixed: 3b2a6ada12210ba15a716d3e302884565dc4b998

rpenido commented:
> This is a weird bug, but if I go to the "Libraries" tab, then click on any library, then click back, the "New Library" button disappears.
> 
> Edit: also, just refreshing the Studio Home makes the buttons disappear?
> 
> Edit: I think this bug is on master too, actually

This seems to fix the issue: 01e817a7bc5c95967991f8a84d4ee0a598b91821

rpenido commented on src/header/Header.tsx:
```diff
@@ -1,83 +1,76 @@
-  ];
-  const outlineLink = `${studioBaseUrl}/course/${courseId}`;
+  ] : [];
+  const outlineLink = !isLibrary ? `${studioBaseUrl}/course/${contentId}` : `/course-authoring/library/${contentId}`;
```
Done: 9937a590d22738b3647933b5d9fe8df49d2fe39c

rpenido commented:
Thank you for the reviews @bradenmacdonald and @xitij2000!
I addressed the issues and added a comment about the search button in the head bar.

This is ready for another round.

bradenmacdonald commented:
Nice work on addressing those comments, @rpenido. @xitij2000 did you want to look at this again before we merge it?

openedx-webhooks commented:
@rpenido 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/frontend-app-course-authoring&entry.867055334=https://github.com/openedx/frontend-app-course-authoring/pull/1076&entry.1484655318=https://github.com/rpenido&entry.752974735=2024-06-06+18:49&entry.1917517419=2024-07-10+12:20&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
